tests/test_qaoa.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the :mod:`pennylane.qaoa` submodule.
+"""
+import pytest
+import networkx as nx
+import numpy as np
+import pennylane as qml
+from pennylane import qaoa
+
+
+#####################################################
+
+graph = nx.Graph()
+graph.add_nodes_from([0, 1, 2])
+graph.add_edges_from([(0, 1), (1, 2)])
+
+
+class TestMixerHamiltonians:
+    """Tests that the mixer Hamiltonians are being generated correctly"""
+
+    def test_x_mixer_output(self):
+        """Tests that the output of the Pauli-X mixer is correct"""
+
+        num_qubits = 4
+        wires = range(num_qubits)
+
+        mixer_hamiltonian = qaoa.x_mixer(wires)
+
+        mixer_coeffs = mixer_hamiltonian.coeffs
+        mixer_ops = [i.name for i in mixer_hamiltonian.ops]
+        mixer_wires = [i.wires[0] for i in mixer_hamiltonian.ops]
+
+        assert (
+            mixer_coeffs == [1, 1, 1, 1]
+            and mixer_ops == ["PauliX", "PauliX", "PauliX", "PauliX"]
+            and mixer_wires == [0, 1, 2, 3]
+        )
+
+    def test_xy_mixer_type_error(self):
+        """Tests that the XY mixer throws the correct error"""
+
+        graph = 12
+
+        with pytest.raises(ValueError) as info:
+            output = qaoa.xy_mixer(graph)
+
+        assert "Input graph must be a networkx.Graph object or Iterable, got int" in str(info.value)
+
+    @pytest.mark.parametrize(
+        ("graph", "target_hamiltonian"),
+        [
+            (
+                [(0, 1), (1, 2), (2, 3)],
```

Comment:
There is an edge case that is missing; if the user passes a list of iterables where the node indices are not strictly increasing:

```python
[(0, 4), (3, 4), (2, 1), (2, 0)]
```
etc.